navis/nbl/smat.py

```diff
@@ -307,15 +350,15 @@ def _pick_nonmatching_pairs(self, n_matching_qual_vals):
         # means that we can parallelise the later step more effectively.
         # Slowdowns here are practically meaningless
         # because of how long distdot calculation will take
-        all_nonmatching_pairs = list(self._yield_nonmatching_pairs())
         nonmatching_pairs = []
         n_nonmatching_qual_vals = 0
         rng = np.random.default_rng(self.seed)
-        while n_nonmatching_qual_vals < n_matching_qual_vals:
-            idx = rng.integers(0, len(all_nonmatching_pairs))
-            nonmatching_pair = all_nonmatching_pairs.pop(idx)
+        for nonmatching_pair in self._yield_nonmatching_pairs(rng=rng):
```

Comment:
This is the change matching above comment re creating only as many non-matching pairs as we absolutely need.